### PR TITLE
Writable windows_sync_objects table

### PR DESCRIPTION
**Pull request by alessandrogario**
This is a port of the old windows_sync_objects table to writable extension tables

### CODE REVIEW

File: windows_sync_objects/src/objectmanager.cpp [hunk 1]
```diff
@@ -0,0 +1,397 @@
...
+      auto error_code = RtlNtStatusToDosError(status);
+      VLOG(1) << "Directory query failed with the following error: "
+              << GetErrorMessage(error_code);
+      break;
+    }
+
+    auto directory_entries =
+        reinterpret_cast<const OBJECT_DIRECTORY_INFORMATION*>(buffer.data());
+
+    for (std::size_t i = 0U; directory_entries[i].Name.Length != 0; i++) {
```
Comment (yying) [inline windows_sync_objects/src/objectmanager.cpp]: Is it possible for one of the returned directory entry names to be an empty string? 

From `UNICODE_STRING` docs at MSDN:

>Windows 7, Windows Server 2008, Windows Vista, Windows Server 2003 and Windows XP:  When the Length structure member is zero and the MaximumLength structure member is 1, the Buffer structure member can be an empty string or contain solely a null character. This behavior changed beginning with Windows Server 2008 R2 and Windows 7 with SP1.

**Comment by alessandrogario** [comment]
Hey @yying

I have used the NtCreate* APIs in order to let the user specify the full path for the object. I was hoping this would improve the capabilities of the table (like acquiring mutants in the context of other users - but I'm not 100% sure it can be done). Do you think this is not necessary?

File: windows_sync_objects/src/objectmanager.cpp [hunk 2]
```diff
@@ -0,0 +1,397 @@
...
+    std::memset(buffer.data(), static_cast<int>(buffer.size()), 0);
+
+    auto status = NtQueryDirectoryObject(directory,
+                                         buffer.data(),
+                                         static_cast<ULONG>(buffer.size()),
+                                         FALSE,
+                                         entry_list.empty() ? TRUE : FALSE,
+                                         &entry_index,
+                                         &response_size);
+    if (NT_ERROR(status)) {
```
[the remark this answers is not in the document]
Comment (alessandrogario) [reply windows_sync_objects/src/objectmanager.cpp]: I have no idea why it was returning ULONG! I fixed it now!

File: windows_sync_objects/src/objectmanager.cpp (hunk 1 above)
Comment (alessandrogario) [reply windows_sync_objects/src/objectmanager.cpp]: I'm now checking all fields to make sure

**Pull request merged**